pkg/sources/client/azure/servicebus/client.go

```diff
@@ -0,0 +1,145 @@
+func NewClientGetter(sg NamespacedSecretsGetter) *ClientGetterWithSecretGetter {
+	return &ClientGetterWithSecretGetter{
+		sg: sg,
+	}
+}
+
+// NamespacedSecretsGetter returns a SecretInterface for the given namespace.
+type NamespacedSecretsGetter func(namespace string) coreclientv1.SecretInterface
+
+// ClientGetterWithSecretGetter gets Azure clients using static credentials
+// retrieved using a Secret getter.
+type ClientGetterWithSecretGetter struct {
+	sg NamespacedSecretsGetter
+}
+
+// ClientGetterWithSecretGetter implements ClientGetter.
+var _ ClientGetter = (*ClientGetterWithSecretGetter)(nil)
+
+// Get implements ClientGetter.
+func (g *ClientGetterWithSecretGetter) Get(src *v1alpha1.AzureServiceBusTopicSource) (*Namespace, error) {
+	azureEnv := &azure.PublicCloud
+
+	secretCli := g.sg(src.Namespace)
+
+	var tokenProvider amqpauth.TokenProvider
+
+	switch {
+	case src.Spec.Auth.ServicePrincipal != nil:
+		tpCtor, err := auth.NewAADTokenProviderCtor(secretCli, src.Spec.Auth.ServicePrincipal)
+		if err != nil {
+			return nil, fmt.Errorf("obtaining constructor for an AAD token provider: %w", err)
+		}
+
+		aadTokenProvider, err := tpCtor(
+			aad.JWTProviderWithAzureEnvironment(azureEnv),
+			// aad.NewJWTProvider configures the TokenProvider to
+			// target Event Hubs by default, but this is not what
+			// we want here.
+			aad.JWTProviderWithResourceURI(azureEnv.ResourceIdentifiers.ServiceBus),
+		)
+		if err != nil {
+			// TODO(antoine): Fix returned error upstream.
+			//
+			// As of v3.2.1, the Azure AMQP module attempts to refresh AAD tokens prior to returning a
+			// TokenProvider but, in case of failure, returns an untyped and unformatted error (with
+			// typos!): https://github.com/Azure/azure-amqp-common-go/blob/v3.2.1/aad/jwt.go#L160-L162
+			//
+			// Example:
+			//   "failed to refersh token: &{{{  0 0 0  } 0xc001d4a240 {{https  <nil> login.microsoftonline.com
+			//   /00000000-0000-0000-0000-000000000000  false   } {https  <nil> login.microsoftonline.com
+			//   /00000000-0000-0000-0000-000000000000/oauth2/authorize  false api-version=1.0  } {https
+			//   <nil> login.microsoftonline.com /00000000-0000-0000-0000-000000000000/oauth2/token
+			//   false api-version=1.0  } {https  <nil>
+			//   login.microsoftonline.com /00000000-0000-0000-0000-000000000000/oauth2/devicecode  false
+			//   api-version=1.0  }} 00000000-0000-0000-0000-000000000000 https://servicebus.azure.net/ true
+			//   300000000000} 0xc001f59380 0xc0029ac8a0 <nil> [] 0}"
+			//
+			// This error contains unique elements such as pointer addresses, and therefore needs to be
+			// sanitized to avoid causing an infinite loop of reconciliations when written to an API
+			// object's status.
+			if strings.Contains(err.Error(), "failed to refersh token") { // (!) the typo is not a mistake
+				err = errors.New("failed to refresh Service Principal token. The provided secret is " +
+					"either invalid or expired")
+			}
```

Comment:
This whole error handling based on string parsing is disgusting and won't be maitainable over time.
We need a better way to figure what actually happened when an error occurs, and this SDK isn't mature enough to provide us that answer.